### DISCUSSION TITLE
Issue with updating `ui.chart` of type `stockChart`

### DISCUSSION THREAD

**Discussion by gengyudong** [zauberzeug/nicegui#948] [Q&A] [answered]
### Question

I'm not sure if this is an issue in my code or an issue with ui.update() or an issue with highcharts library. When I cleared the series list and tried to append 4 other series into the chart, only 3 got appended, with the last one being left out. Additionally, when I click the "append" button twice, random series (with random names - 'series4') got appended to the chart as well (attached a screenshot of the chart below), which I have no clue why and how it happened. Appreciate any advice or comment for this occurrence. 

```python
data1 = {'name': 'alice', 'data': [31, 148, 33, 68, 75, 43]}
data2 = {'name': 'bob', 'data': [3, 28, 7, 5, 8, 3]}
data3 = {'name': 'john', 'data': [1, 4, 0, 8, 9, 0]}  
data4 = {'name': 'tom', 'data': [1, 13, 2, 4, 1, 0]}    

chart = ui.chart({
        'chart': {
            'type': 'spline',
            'zoomType': 'xy',
            'backgroundColor': 'rgba(0,0,0,0)',
        },
        'legend': {
            'enabled': True,
            'layout': 'vertical',
            'align': 'right',
            'verticalAlign': 'middle',
        },
        'rangeSelector': {
            'allButtonsEnabled': True,
                'buttons': [{
                    'type': 'month',
                    'count': 3,
                    'text': 'Day',
                    'dataGrouping': {
                        'forced': True,
                        'units': [['day', [1]]]
                    }
                }, {
                    'type': 'year',
                    'count': 1,
                    'text': 'Week',
                    'dataGrouping': {
                        'forced': True,
                        'units': [['week', [1]]]
                    }
                }, {
                    'type': 'all',
                    'text': 'Month',
                    'count': 1,
                    'dataGrouping': {
                        'forced': True,
                        'units': [['month', [1]]]
                    }
                }],
                'buttonTheme': {
                    'width': 60
                },
                'inputStyle': {
                    'color': '#CED5DF'
                },
                'selected': 0,
        },
        'navigator': {
            'enabled': True
        },
        'scrollbar': {
            'enabled': False
        },
        'plotOptions': {
            'series': {
                'dataLabels': {'enabled': True},
                'label': {'connectorAllowed': False},
                'pointStart': 1577808000,
                'pointInterval': 24 * 3600 * 1000,
                'dataGrouping': {
                    'approximation': 'sum'
                }
            }
        },
        'series': [data1]
    }, type = 'stockChart', extras = ['stock'])

def update():
    chart.options['series'].append(data2)
    chart.options['series'].append(data3)
    chart.options['series'].append(data4)
    chart.update()
    
def revert():
    chart.options['series'].clear()
    chart.update()
  
ui.button('update', on_click=update)
ui.button('revert', on_click=revert)

ui.run(port = 8088)
```

<img width="333" alt="image" src="https://github.com/zauberzeug/nicegui/assets/76908945/26ed55d6-89da-4145-91bb-c5c26e804b3a">

**Comment by falkoschindler** [comment]
Very interesting discovery, @gengyudong! This looks like a bug. Here is a minimal reproduction:

```py
chart = ui.chart({'legend': {'enabled': True}, 'series': []}, type='stockChart', extras=['stock'])
ui.button('update', on_click=lambda: (
    chart.options['series'].extend([{'name': 'alice'}, {'name': 'bob'}]),
    chart.update(),
))
```

Alice _and_ Bob should be appear in the legend, but only Alice does.
It only happens when creating a `stockChart`.

**Comment by falkoschindler** [comment]
It seems to be caused by these lines:
https://github.com/zauberzeug/nicegui/blob/d0ab338e3bf1eb2cc3626d126ddf0e3f8db4d1a8/nicegui/elements/chart.js#L21-L23

Because `chart.update` does not work for adding and removing series, we add and remove them beforehand in two while loops. But somehow the first `addSeries` causes the length to jump from 0 to 2. So the second series is not added. But I still don't know why this happens, why it happens only for stock charts, and - most importantly - what to do about it.

**Comment by falkoschindler** [comment, marked as answer]
Ok, I got it. PR https://github.com/zauberzeug/nicegui/pull/951 should fix it.

**Reply by gengyudong** [reply]
That's a quick fix! Thank you very much!

**Reply by gengyudong** [reply]
Hi @falkoschindler I am still facing the issue of only having 1 series being added instead of 2. I have updated my chart.js file with the updated one in PR #951, is there anything else that I should be doing to reflect the fix on my end?

**Reply by falkoschindler** [reply]
@gengyudong Your browser might have cached the old chart.js file. Could you, please, clear the cache and try again?
